### PR TITLE
Update the IBufferWriter contract definition documentation to be clearer

### PR DESCRIPTION
**Pull request by ahsonkhan**
Based on recent discussions:
- https://github.com/dotnet/corefx/issues/34259
- https://github.com/dotnet/corefx/pull/35186

cc @pakrym, @stephentoub, @davidfowl, @AArnott, @benaadams, @KrzysztofCwalina, @terrajobst

### CODE REVIEW

File: src/System.Memory/src/System/Buffers/IBufferWriter.cs [hunk 1]
```diff
@@ -5,25 +5,48 @@
 namespace System.Buffers
 {
     /// <summary>
-    /// Represents a <typeparam name="T"/> sink
+    /// Represents a <typeparam name="T"/> output sink into which data can be written.
```
Comment (ahsonkhan) [inline src/System.Memory/src/System/Buffers/IBufferWriter.cs]: Should I add, `synchronously`?

**Comment by ahsonkhan** [comment]
Changes to the docs repo:
https://github.com/dotnet/dotnet-api-docs/pull/1949

cc @mairaw

File: src/System.Memory/src/System/Buffers/IBufferWriter.cs (hunk 1 above)
Comment (KrzysztofCwalina) [reply src/System.Memory/src/System/Buffers/IBufferWriter.cs]: T is not a sink or output. It's the coding unity of the data being written.

Comment (ahsonkhan) [reply src/System.Memory/src/System/Buffers/IBufferWriter.cs]: `Represents an output sink into which <typeparam name="T"/> data can be written`?

Comment (KrzysztofCwalina) [reply src/System.Memory/src/System/Buffers/IBufferWriter.cs]: But this doc comment is for type parameter T, not for the whole type IBufferWriter<T>, isn't it?

Comment (ahsonkhan) [reply src/System.Memory/src/System/Buffers/IBufferWriter.cs]: No, this doc comment is for the whole type: `IBufferWriter<T>`

For example see:
https://docs.microsoft.com/en-us/dotnet/api/system.buffers.ibufferwriter-1?view=netcore-2.2
image

Comment (KrzysztofCwalina) [reply src/System.Memory/src/System/Buffers/IBufferWriter.cs]: Ah, sorry. I missed that. Looks good then.

File: src/System.Memory/src/System/Buffers/IBufferWriter.cs [hunk 2]
```diff
@@ -5,25 +5,48 @@
...
+        /// <remarks>
+        /// You must request a new buffer after calling Advance to continue writing more data and cannot write to a previously acquired buffer.
+        /// </remarks>
         void Advance(int count);
 
         /// <summary>
-        /// Requests the <see cref="Memory{T}"/> that is at least <paramref name="sizeHint"/> in size if possible, otherwise returns maximum available memory.
-        /// If <paramref name="sizeHint"/> is equal to <code>0</code>, currently available memory would get returned.
+        /// Returns a <see cref="Memory{T}"/> to write to that is at least the requested length (specified by <paramref name="sizeHint"/>).
+        /// If no <paramref name="sizeHint"/> is provided (or it's equal to <code>0</code>), some, non-empty, buffer is returned.
```
Comment (stephentoub) [inline src/System.Memory/src/System/Buffers/IBufferWriter.cs]: Nit: remove the commas around "non-empty"

File: src/System.Memory/src/System/Buffers/IBufferWriter.cs [hunk 3]
```diff
@@ -5,25 +5,48 @@
...
+        /// <remarks>
+        /// You must request a new buffer after calling Advance to continue writing more data and cannot write to a previously acquired buffer.
+        /// </remarks>
         Memory<T> GetMemory(int sizeHint = 0);
 
         /// <summary>
-        /// Requests the <see cref="Span{T}"/> that is at least <paramref name="sizeHint"/> in size if possible, otherwise returns maximum available memory.
-        /// If <paramref name="sizeHint"/> is equal to <code>0</code>, currently available memory would get returned.
+        /// Returns a <see cref="Span{T}"/> to write to that is at least the requested length (specified by <paramref name="sizeHint"/>).
+        /// If no <paramref name="sizeHint"/> is provided (or it's equal to <code>0</code>), some, non-empty, buffer is returned.
```
Comment (stephentoub) [inline src/System.Memory/src/System/Buffers/IBufferWriter.cs]: Same nit

**Pull request merged**